src/main/java/com/ibm/cloud/sdk/core/security/IamToken.java

```diff
@@ -51,28 +52,51 @@ public Long getExpiration() {
     return expiration;
   }
 
+  private IamToken() {
+    if (getExpiresIn() != null && getExpiration() != null) {
+      Double fractionOfTimeToLive = 0.8;
+      Long timeToLive = getExpiresIn();
+      Long expirationTime = getExpiration();
+      this.refreshTime = expirationTime - (long) (timeToLive * (1.0 - fractionOfTimeToLive));
+    }
+  }
+
   /**
-   * Check if currently stored access token is valid.
+   * Returns true iff currently stored access token should be refreshed.
    *
-   * Using a buffer to prevent the edge case of the
-   * token expiring before the request could be made.
+   * This method uses a buffer to prevent the edge case of the
+   * token expiring before the request could be made, so this method will return true
+   * if the current time is within that buffer.
    *
    * The buffer will be a fraction of the total TTL. Using 80%.
    *
-   * @return true iff the current access token is valid and not expired
+   * This method also updates the expiration time if it determines the token needs refreshed to prevent other
+   * threads from making multiple refresh calls.
+   *
+   * @return true iff the current access token is invalid or past the expiration buffer
    */
   @Override
-  public boolean isTokenValid() {
-    if (getAccessToken() == null || getExpiresIn() == null || getExpiration() == null) {
-      return false;
+  public synchronized boolean needsRefresh() {
+    long currentTime = System.currentTimeMillis() / 1000;
+
```

Comment:
Add the logic to calculate `refreshTime` here if `this.refreshTime == null`.  If you always set `refreshTime` to a non-null value, it will only be executed once -- so still efficient.